internal/scheduler/ingester.go

```diff
@@ -0,0 +1,777 @@
+	}
+}
+
+// Batch of reprioritisations.
+// PrioritiesByJobSet should always be applied before PrioritiesByJob.
+// ReprioritisationBatch
+// To ensure changes are applied in the correct order.
+type ReprioritisationBatch struct {
+	PrioritiesByJobSet map[string]int64
+	PrioritiesByJob    map[uuid.UUID]int64
+}
+
+func NewReprioritisationBatch() *ReprioritisationBatch {
+	return &ReprioritisationBatch{
+		PrioritiesByJobSet: make(map[string]int64),
+		PrioritiesByJob:    make(map[uuid.UUID]int64),
+	}
+}
+
+// ShouldWrite returns true if this batch should be written into postgres.
+// TODO: Update
+func (srv *Ingester) ShouldWrite(batch *Batch) bool {
+	if batch == nil {
+		return false
+	}
+	if time.Since(batch.CreatedAt) > srv.MaxWriteInterval {
+		return true
+	}
+	if len(batch.MessageIds) > srv.MaxWriteRecords {
+		return true
+	}
+	return false
+}
+
+// Run the ingester until experiencing an unrecoverable error.
+func (srv *Ingester) Run(ctx context.Context) error {
+
+	// Get the configured logger, or the standard logger if none is provided.
+	var log *logrus.Entry
+	if srv.Logger != nil {
+		log = srv.Logger.WithField("service", "SchedulerIngester")
+	} else {
+		log = logrus.StandardLogger().WithField("service", "SchedulerIngester")
+	}
+
+	log.Info("service started")
+
+	// Create a processing pipeline
+	// receive from pulsar -> batch messages -> write to postgres and ack messages.
+	for {
+
+		// Scheduler ingester.
+		consumer, err := srv.PulsarClient.Subscribe(srv.ConsumerOptions)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		defer consumer.Close()
+		srv.consumer = consumer
+
+		// All services run within an errgroup.
+		g, ctx := errgroup.WithContext(ctx)
+		ctx = ctxlogrus.ToContext(ctx, log)
+
+		// Receive Pulsar messages asynchronously.
+		pulsarToChannel := pulsarutils.NewPulsarToChannel(srv.consumer)
+		g.Go(func() error { return pulsarToChannel.Run(ctx) })
+
+		// Batch messages for writing into postgres.
+		// Batches are forwarded on batchChannel.
+		batchChannel := make(chan *Batch)
+		g.Go(func() error { return srv.ProcessMessages(ctx, pulsarToChannel.C, batchChannel) })
+
+		// Write batches to postgres and, for each batch, ack all pulsar messages the batch was made up of.
+		g.Go(func() error { return srv.ProcessBatches(ctx, batchChannel) })
+
+		// Run pipeline until any error.
+		err = g.Wait()
+		// TODO: Detect recoverable errors.
+		if err != nil {
+			// Unrecoverable error; exit.
+			return err
+		}
+	}
+}
+
+func (srv *Ingester) ProcessMessages(ctx context.Context, messageChannel <-chan pulsar.Message, batchChannel chan<- *Batch) error {
+	log := ctxlogrus.Extract(ctx)
+
+	// In-progress batches.
+	batches := make([]*Batch, 0)
+
+	// Ticker to trigger periodically writing to postgres.
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	// Run until ctx is canceled.
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case msg := <-messageChannel:
```

Comment:
If I'm honest I'm not a massive fan of this because it mixes in the batching logic with other service-specific logic.    Now that we have generics we should be able to write a generic batching function and then write service-specific functions to operate on either a batch or single element as required.